### PR TITLE
Initial implementation of structured logging in cloud backend

### PR DESCRIPTION
**Pull request by sebasslash**
Currently, the Terraform `cloud` backend has no way of rendering streamed JSON logs, which resulted in structured run output being disabled for CLI-driven runs. These are the changes that enable the `cloud` backend to consume structured logs and make use of the new plan renderer. This will allow CLI-driven runs to enable the structured output in Terraform Cloud.

What we want is to convert JSON logs into human-readable format and a plan renderer that can convert the JSON plan output to human-readable as well. The renderer was introduced in https://github.com/hashicorp/terraform/pull/32496. The cloud backend will call this renderer to generate the plan output instead of relying on the streamed logs for plan output as it did before. 

Furthermore, these changes to the `cloud` backend will maintain backwards compatibility with any streamed logs that are already in human-readable format. 

In short, the goal here is that you don't notice a change in your current Terraform CLI experience (you will see plaintext logs no matter what) meanwhile allowing for closer parity with the Terraform Cloud experience. 

## Target Release

1.4.0

### ENHANCEMENTS

 

-  The `cloud` backend now renders structured logs and plan JSON output in human-readable format, allowing for CLI-driven TFC workspaces to enable structured run output. 

### Testing (Internal Employees)
Only internal employees can test as of this moment (sorry :sad:) since the related TFC changes have not been released. Please see the TFC PR titled: `Enable SRO for CLI-Driven runs` for more information on how to test.

### CODE REVIEW

File: internal/cloud/backend_apply.go [hunk 1]
```diff
@@ -151,41 +153,69 @@ func (b *Cloud) opApply(stopCtx, cancelCtx context.Context, op *backend.Operatio
...
 
 			if next || len(line) > 0 {
-				b.CLI.Output(b.Colorize().Color(string(line)))
+				log := &jsonformat.JSONLog{}
+				if err := json.Unmarshal(line, log); err != nil || log == nil {
+					// If we can not parse the line as JSON, we will simply
+					// print the line. This maintains backwards compatibility for
+					// users who do not wish to enable structured output in their
+					// workspace.
+					b.CLI.Output(b.Colorize().Color(string(line)))
```
Comment (alisdair) [inline internal/cloud/backend_apply.go:203]: I don't think we should be colorizing these lines.

File: internal/cloud/backend_plan.go [hunk 2]
```diff
@@ -373,6 +353,128 @@ in order to capture the filesystem context the remote workspace expects:
...
+			}
+
+			if next || len(line) > 0 {
+				log := &jsonformat.JSONLog{}
+				if err := json.Unmarshal(line, log); err != nil || log == nil {
+					// If we can not parse the line as JSON, we will simply
+					// print the line. This maintains backwards compatibility for
+					// users who do not wish to enable structured output in their
+					// workspace.
+					b.CLI.Output(b.Colorize().Color(string(line)))
```
Comment (alisdair) [inline internal/cloud/backend_plan.go:400]: Same note about colorization as above. We should never be receiving color codes (e.g. `[red]`) from a cloud text stream, only plain text with possible ANSI color codes.

File: internal/cloud/backend_plan.go [hunk 3]
```diff
@@ -373,6 +353,128 @@ in order to capture the filesystem context the remote workspace expects:
...
+					continue
+				}
+
+				// We will ignore outputs logs during the plan phase
+				if log.Type == jsonformat.LogOutputs {
+					continue
+				}
+
+				// We'll defer any log during a plan operation that is not
+				// plan output or outputs logs
```
Comment (alisdair) [inline internal/cloud/backend_plan.go:410]: Interesting. What are some examples of logs we should defer here?

File: internal/command/jsonformat/renderer.go [hunk 4]
```diff
@@ -27,6 +30,27 @@ const (
...
 )
 
+type JSONLogType string
+
+type JSONLog struct {
+	Message    string                `json:"@message"`
+	Type       JSONLogType           `json:"type"`
+	Diagnostic *viewsjson.Diagnostic `json:"diagnostic"`
+	Outputs    viewsjson.Outputs     `json:"outputs"`
+}
```
Comment (alisdair) [inline internal/command/jsonformat/renderer.go:40]: This seems pragmatic for now.

For future: did you consider using something like mapstructure (or a similar technique) to inspect the type and decode into a detailed struct? This would allow us to more closely unify the UI output for local and remote operations, if both are rendering data from the same source. It would also allow us to more usefully colorize log messages where appropriate.

File: internal/command/jsonformat/renderer.go [hunk 5]
```diff
@@ -240,6 +264,49 @@ func (r Renderer) RenderHumanPlan(plan Plan, mode plans.Mode, opts ...RendererOp
 	}
 }
 
+func (r Renderer) RenderLog(log *JSONLog) error {
+	switch log.Type {
+	case LogApplyStart, LogApplyComplete, LogRefreshStart:
+		msg := fmt.Sprintf("[bold]%s[reset]", log.Message)
+		r.Streams.Println(r.Colorize.Color(msg))
```
Comment (alisdair) [inline internal/command/jsonformat/renderer.go:271]:
```suggestion
		r.Streams.Printf(r.Colorize.Color("[bold]%s[reset]"), log.Message)
```

As a rule, we must not colorize an interpolated string, as it means that any instances of color codes (e.g. `[red]`) inside user-defined content will be incorrectly rendered. Instead, we can colorize the format string, then pass that result to `Printf`.

File: internal/command/jsonformat/renderer.go [hunk 6]
```diff
@@ -240,6 +264,49 @@ func (r Renderer) RenderHumanPlan(plan Plan, mode plans.Mode, opts ...RendererOp
...
+
+				msg := fmt.Sprintf("%s = %s", name, outputStr)
+				r.Streams.Println(r.Colorize.Color(msg))
+			}
+		}
+
+	case LogChangeSummary:
+		// We will only render the apply change summary since the renderer
+		// generates a plan change summary for us
+		if !strings.Contains(log.Message, "Plan") {
```
Comment (alisdair) [inline internal/command/jsonformat/renderer.go:301]: This feels like a hack. Is this necessary, or can we prevent calling `RenderLog` with this type when we're in plan mode?

File: internal/command/jsonformat/renderer.go [hunk 7]
```diff
@@ -240,6 +264,49 @@ func (r Renderer) RenderHumanPlan(plan Plan, mode plans.Mode, opts ...RendererOp
...
+			msg := fmt.Sprintf("[bold][green]%s[reset]", log.Message)
+			r.Streams.Println("\n" + r.Colorize.Color(msg) + "\n")
```
Comment (alisdair) [inline internal/command/jsonformat/renderer.go:303]: Same comment as above about colorizing format strings.

File: internal/command/jsonformat/renderer.go [hunk 8]
```diff
@@ -240,6 +264,49 @@ func (r Renderer) RenderHumanPlan(plan Plan, mode plans.Mode, opts ...RendererOp
...
+				}
+
+				outputDiff := change.ComputeDiffForType(ctype)
+				outputStr := outputDiff.RenderHuman(0, computed.RenderHumanOpts{
+					Colorize:              r.Colorize,
+					ShowUnchangedChildren: true,
+				})
+
+				msg := fmt.Sprintf("%s = %s", name, outputStr)
+				r.Streams.Println(r.Colorize.Color(msg))
```
Comment (alisdair) [inline internal/command/jsonformat/renderer.go:294]: I don't believe we should be colorizing this at all.

**Review by alisdair — COMMENTED** [review]
Some initial thoughts while in draft!

File: internal/command/jsonformat/renderer.go [hunk 9]
```diff
@@ -240,6 +264,49 @@ func (r Renderer) RenderHumanPlan(plan Plan, mode plans.Mode, opts ...RendererOp
...
+		}
+
+	case LogChangeSummary:
+		// We will only render the apply change summary since the renderer
+		// generates a plan change summary for us
+		if !strings.Contains(log.Message, "Plan") {
+			msg := fmt.Sprintf("[bold][green]%s[reset]", log.Message)
+			r.Streams.Println("\n" + r.Colorize.Color(msg) + "\n")
+		}
+	}
```
Comment (alisdair) [inline internal/command/jsonformat/renderer.go:305]: This switch statement has no `default` case, which means that anything not explicitly handled here will be dropped. This is not forwards-compatible with Terraform versions which emit new logs, as they won't be rendered if the client version is older. That was the purpose of the `@message` field: allowing older clients to have a fallback output for new information.

To correct this I think we'll need a stop-list of messages we _don't_ want to render, and a default case which just prints out the `@message` field for any unrecognized types.

File: internal/cloud/backend_plan.go (hunk 3 above)
Comment (sebasslash) [reply internal/cloud/backend_plan.go:410]: This initially was for deferring the change summary log. Given the new plan renderer does this for us, I'll remove this logic.

File: internal/command/jsonformat/renderer.go (hunk 4 above)
Comment (sebasslash) [reply internal/command/jsonformat/renderer.go:40]: I did not consider that. I was wondering if you could explain this in more detail -- how would mapstructure differ from `viewjson.Outputs` ?

Comment (sebasslash) [reply internal/command/jsonformat/renderer.go:40]: I am unfamiliar with mapstructure so I did not consider it. Is this in reference to the entire `JSONLog` struct or the `Outputs` field specifically? Wondering what advantages mapstructure would give us -- reading through the [README](https://github.com/mitchellh/mapstructure) at first glance seems like the library is a sensible option if we can't make any guarantees about the schema for logs.

Comment (alisdair) [reply internal/command/jsonformat/renderer.go:40]: mapstructure is just an example of a technique for decoding differently-structured JSON messages into the appropriate structs. Looking through the Terraform code base, it seems like we don't tend to use mapstructure itself in modern code.

The advantage we'd get from doing this at all is that we would be able to decode different messages into different structures, which we could then render based upon that decoded data. While we could have a giga-struct which is the overlay of all possible structs, that's messy, confusing, and may not be possible if there are naming conflicts.

This isn't something we need to address now, by any means! As far as I can see, with this solution is a very reasonable tradeoff for now. I'd be happy to talk more or jam on this a bit as a follow-up, if you like.

File: internal/cloud/backend_common.go [hunk 10]
```diff
@@ -538,3 +543,38 @@ func (b *Cloud) confirm(stopCtx context.Context, op *backend.Operation, opts *te
...
+
+	req.Header.Set("Authorization", "Bearer "+token)
+	req.Header.Set("Accept", "application/json")
+
+	p := &jsonformat.Plan{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return p, err
+	}
+
```
Comment (jbardin) [inline internal/cloud/backend_common.go:574]: not going to cause any major problems in a CLI, but we should still `defer resp.Body.Close()` to close the connection.

File: internal/cloud/backend_apply.go [hunk 11]
```diff
@@ -151,41 +153,69 @@ func (b *Cloud) opApply(stopCtx, cancelCtx context.Context, op *backend.Operatio
...
+			// (which contain version and os/arch information), we therefore skip to prevent duplicate output.
 			if skip < 3 {
 				skip++
 				continue
 			}
 
 			if next || len(line) > 0 {
-				b.CLI.Output(b.Colorize().Color(string(line)))
+				log := &jsonformat.JSONLog{}
+				if err := json.Unmarshal(line, log); err != nil || log == nil {
```
Comment (brandonc) [inline internal/cloud/backend_apply.go:198]: 💅 will log ever be `nil` ? I think not since you are assigning it the address of this struct

File: internal/cloud/backend_plan.go [hunk 12]
```diff
@@ -373,6 +354,121 @@ in order to capture the filesystem context the remote workspace expects:
...
 
+// renderPlanLogs reads the streamed plan JSON logs and calls the JSON Plan renderer (jsonformat.RenderPlan) to
+// render the plan output. The plan output is fetched from the redacted output endpoint.
+func (b *Cloud) renderPlanLogs(ctx context.Context, op *backend.Operation, run *tfe.Run) error {
+	logs, err := b.client.Plans.Logs(ctx, run.Plan.ID)
+	if err != nil {
+		return err
+	}
+
+	reader := bufio.NewReaderSize(logs, 64*1024)
```
Comment (brandonc) [inline internal/cloud/backend_plan.go:365]: 💅 You only need to allocate this if b.CLI != nil

File: internal/cloud/backend_plan.go [hunk 13]
```diff
@@ -373,6 +354,121 @@ in order to capture the filesystem context the remote workspace expects:
...
+					}
+					next = false
+				}
+
+				line = append(line, l...)
+			}
+
+			if next || len(line) > 0 {
+				log := &jsonformat.JSONLog{}
+				if err := json.Unmarshal(line, log); err != nil || log == nil {
```
Comment (brandonc) [inline internal/cloud/backend_plan.go:386]: 💅 log is never `nil` (similar to above)

File: internal/cloud/backend_plan.go [hunk 14]
```diff
@@ -373,6 +354,121 @@ in order to capture the filesystem context the remote workspace expects:
...
+
+	return nil
+}
+
+func satisfiesStructuredOutputTerraformVersion(workspace *tfe.Workspace) bool {
+	if workspace == nil {
+		return false
+	}
+
+	workspaceVersion, err := version.NewVersion(workspace.TerraformVersion)
```
Comment (brandonc) [inline internal/cloud/backend_plan.go:458]: `workspace.TerraformVersion` can also be a [version constraint](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/workspaces#update-a-workspace), and not a version string, which would make this function always return false (not certain that is desirable)

This go playground illustrates the issue of trying to compare a constraint to a version. We https://go.dev/play/p/YUvyfNzsjT0

Maybe the error behavior is fine, but we should do some debug logging in this function?

File: internal/cloud/backend_plan.go [hunk 15]
```diff
@@ -373,6 +354,121 @@ in order to capture the filesystem context the remote workspace expects:
...
+		run.Status == tfe.RunDiscarded {
+		// We won't return an error here since we need to resume the logic that
+		// follows after rendering the logs (run tasks, cost estimation, etc.)
+		return nil
+	}
+
+	// Given CLI-driven runs will silently disable SRO if the version is not supported,
+	// we need to ensure the workspace Terraform version is >=1.4.0 and has
+	// structured output enabled.
+	if satisfiesStructuredOutputTerraformVersion(run.Workspace) && b.renderer != nil {
```
Comment (brandonc) [inline internal/cloud/backend_plan.go:435]: At a high level, what happens if satisfiesStructuredOutputTerraformVersion is false? It looks like no error is returned and no plan is rendered.

**Review by brandonc — COMMENTED** [review]
Mostly just style feedback, but the more serious concern is about version constraints set on workspace terraform-version: What happens? Can you help me understand why it's important to evaluate a workspace terraform version from the CLI?

**Comment by sebasslash** [comment]
Thanks for this very thorough review :+1:. I will actually remove the codepath on checking the workspace Terraform version. This had been done prior to the run source header check being altered -- we don't need to make this check.

File: internal/cloud/backend_common.go [hunk 16]
```diff
@@ -538,3 +543,38 @@ func (b *Cloud) confirm(stopCtx context.Context, op *backend.Operation, opts *te
...
+// This method will fetch the redacted plan output and marshal the response into
+// a struct the jsonformat.Renderer expects.
+//
+// Note: Apologies for the lengthy definition, this is a result of not being able to mock receiver methods
+var readRedactedPlan func(context.Context, string, string, string) (*jsonformat.Plan, error) = func(ctx context.Context, hostname, token, planID string) (*jsonformat.Plan, error) {
+	client := retryablehttp.NewClient()
+	client.RetryMax = 10
+	client.Logger = logging.HCLogger()
+
+	u := fmt.Sprintf("https://%s/api/v2/plans/%s/json-output-redacted",
```
[the remark this answers is not in the document]
Comment (sebasslash) [reply internal/cloud/backend_common.go]: Done

File: internal/cloud/backend_common.go [hunk 17]
```diff
@@ -538,3 +543,37 @@ func (b *Cloud) confirm(stopCtx context.Context, op *backend.Operation, opts *te
...
+// This method will fetch the redacted plan output and marshal the response into
+// a struct the jsonformat.Renderer expects.
+//
+// Note: Apologies for the lengthy definition, this is a result of not being able to mock receiver methods
+var readRedactedPlan func(context.Context, url.URL, string, string) (*jsonformat.Plan, error) = func(ctx context.Context, baseURL url.URL, token string, planID string) (*jsonformat.Plan, error) {
+	client := retryablehttp.NewClient()
+	client.RetryMax = 10
+	client.Logger = logging.HCLogger()
+
+	u, err := baseURL.Parse(fmt.Sprintf(
```
Comment (liamcervante) [inline internal/cloud/backend_common.go:556]: do we need to check `err` here?

File: internal/cloud/backend_common.go [hunk 18]
```diff
@@ -538,3 +543,37 @@ func (b *Cloud) confirm(stopCtx context.Context, op *backend.Operation, opts *te
...
+		return nil, err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+token)
+	req.Header.Set("Accept", "application/json")
+
+	p := &jsonformat.Plan{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return p, err
```
Comment (Uk1288) [inline internal/cloud/backend_common.go:570]: should this return `return nil, err`?

**Review by brandonc — APPROVED** [review]
Whoa, it worked!

**Review by liamcervante — APPROVED** [review]
